### PR TITLE
Improve containerd implementation of Garden backend interface methods

### PR DESCRIPTION
**Pull request by deniseyu**
This PR incorporates code review feedback from #4784, refactors out error types, and adds additional edge case handling to the runtime backend. Some interface methods are not yet implemented (`BulkInfo`, `BulkMetrics`, `Lookup`, and `GraceTime`) because they either read properties of a concrete implementation of the Garden `Container` interface, or they need to return this concrete type. That work will be delivered after #4877.

Everything is behind a feature flag that isn't exposed yet in any packaging, so is safe to continuously merge into master.

# Existing Issue

Addresses #4783 .

# Changes proposed in this pull request

* Create backend error types that can wrap client-side errors 
* Use context with timeout for all function calls that cross the backend/client boundary
* Add more error checks for Destroy edge cases
* Add godocs for backend methods

# Contributor Checklist
- [x] Unit tests
- [x] Integration tests (if applicable)

# Reviewer Checklist
- [ ] Code reviewed
- [ ] Tests reviewed
- [ ] PR acceptance performed

### CODE REVIEW

File: worker/backend/backend.go [hunk 1]
```diff
@@ -112,36 +138,41 @@ func (b *Backend) Create(gdnSpec garden.ContainerSpec) (container garden.Contain
...
+		if err != nil {
+			return ClientError{ InnerError: err }
 		}
-
-		return b.client.Destroy(ctx, handle)
 	}
 
-	timeDelimitedContext, _ := context.WithTimeout(ctx, maxTaskKillWaitTime)
-	err = killTasks(timeDelimitedContext, task)
+	err = b.client.Destroy(ctxWithTimeout, handle)
```
Comment (kcmannem) [inline worker/backend/backend.go]: is the reason we switched from killTasks to Destroy to avoid killing gracefully?

Comment (deniseyu) [reply worker/backend/backend.go]: It's just a refactor - we only want to call `killTasks` if we can reference the container without error (line 165). This is kind of a complicated piece of logic that we attempted to explain better in the unit tests -- we moved `killTasks` up, so that Destroy gets in both of these circumstances:

1. A container is found but no Task was ever started inside it (line 161). This means that container creation was somehow interrupted, maybe by a network partition, so we want this container to be cleanly deletable later on by a GC. We fence off the "task does not exist" genre of error in 161.
1. A container is found and its Task was successfully stopped.

In both of these cases we want Destroy to be the next thing that happens. If no errors trigger early returns, then this should be the control flow now!
